multicluster/controllers/multicluster/clusterset_controller.go

```diff
@@ -60,3 +122,179 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.clusterSetID = common.ClusterSetID(clusterSetClaimID)
+	r.clusterSetConfig = clusterSet.DeepCopy()
+
+	return nil
+}
+
+func (r *ClusterSetReconciler) createOrUpdateMultiClusterSetOnLeaderCluster(clusterSet *multiclusterv1alpha1.ClusterSet) {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	if r.inboundLeaderMgr == nil {
+		r.inboundLeaderMgr = internal.NewInboundLeaderClusterManager(r.Client, r.clusterID, clusterSet.GetNamespace(), r.Log)
+	}
+}
+
+func (r *ClusterSetReconciler) updateMultiClusterSetOnMemberCluster(clusterSet *multiclusterv1alpha1.ClusterSet) error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	if r.remoteClusterMgr == nil {
+		r.remoteClusterMgr = internal.NewRemoteClusterManager(r.clusterSetID, r.Log, r.clusterID)
+		go func() {
+			r.Log.Info("Starting remote cluster manager", "clusterSetID", r.clusterSetID)
+			var err error
+			err = r.remoteClusterMgr.Start()
+			if err != nil {
+				r.Log.Error(err, "Error starting remote remote cluster manager")
+			}
+			r.mutex.Lock()
+			r.remoteClusterMgr = nil
+			r.clusterSetID = common.INVALID_CLUSTER_SET_ID
+			r.clusterID = common.INVALID_CLUSTER_ID
+			r.clusterSetConfig = nil
+			r.mutex.Unlock()
+		}()
+	}
+
+	if r.remoteClusterMgr == nil {
+		return nil
+	}
+
+	currentLeaders := r.remoteClusterMgr.GetRemoteClusters()
+	newLeaders := clusterSet.Spec.Leaders
+
+	var addedLeaders []*multiclusterv1alpha1.MemberCluster
+	var removedLeaders map[common.ClusterID]internal.RemoteCluster
+
+	for _, leader := range newLeaders {
+		leaderID := common.ClusterID(leader.ClusterID)
+		_, found := currentLeaders[leaderID]
+		if !found {
+			addedLeaders = append(addedLeaders, leader.DeepCopy())
+		} else {
+			// In the end currentLeaders will only have removed leaders
+			delete(currentLeaders, leaderID)
+		}
+	}
+
+	ch := make(chan error)
+	var wg sync.WaitGroup
+	wg.Add(len(addedLeaders))
+	go func() {
+		wg.Wait()
+		close(ch)
+	}()
+
+	for _, addedLeader := range addedLeaders {
+		clusterID := common.ClusterID(addedLeader.ClusterID)
+		url := addedLeader.Server
+		secretName := addedLeader.Secret
+
+		r.Log.Info("creating remote cluster", "clusterID", clusterID)
+
+		go func(clusterID common.ClusterID, url string, secretName string) {
+			defer wg.Done()
+
+			_, err := internal.NewRemoteCluster(clusterID, r.clusterSetID, url, secretName, r.Scheme,
+				r.Log, r.remoteClusterMgr, clusterSet.Spec.Namespace, clusterSet.GetNamespace())
+			if err != nil {
+				r.Log.Error(err, "Unable to create remote cluster", "ClusterID", clusterID)
+			} else {
+				r.Log.Info("Created", "ClusterID", clusterID)
+			}
+
+			ch <- err
+		}(clusterID, url, secretName)
+	}
+
+	var err error
+	for errFromCh := range ch {
+		if errFromCh != nil {
+			r.Log.Error(errFromCh, "Received error")
+			err = multierr.Append(err, errFromCh)
+		}
+	}
+
+	removedLeaders = currentLeaders
+	for _, remoteCluster := range removedLeaders {
+		removeErr := r.remoteClusterMgr.RemoveRemoteCluster(remoteCluster)
```

Comment:
seems `RemoveRemoteCluster` only add an remove event to channel instead of remove it, it always return `nil` , I don't think you can get a real error here.